### PR TITLE
Add package cache replacement/retention process

### PR DESCRIPTION
**Pull request by desruc**
This PR add the package retention process.

- `ApplyRetention` function added to the package journal 
  - Deletes package files if there is less than 500mb available in the given directory
  - Removes packages from journal
- `ApplyRetention` runs whenever a Download or Apply Delta command is executed
  - This replaces `EnsureDiskHasEnoughFreeSpace` (which just threw an error if there wasn't enough space)
- `OctopusPackageRetentionCacheSizeInMegaBytes` variable support added to allow for a maximum cache size to be specified.

Previously, all package retention was handled by the `Clean` command. This is removed in this [PR](https://github.com/OctopusDeploy/Calamari/pull/797).

### CODE REVIEW

**Comment by desruc** [comment]
Tests will fail at the moment. Waiting on finalized retention algorithm and paths to be set on packages.

File: source/Calamari.Tests/Fixtures/PackageRetention/JournalEntryFixture.cs [hunk 1]
```diff
@@ -128,12 +139,30 @@ public void WhenPackageIsRegisteredAndDeregistered_ThenUsageIsStillRecorded()
...
+            var deploymentOne = new ServerTaskId("Deployment-1");
+    
+            var journal = GetPackageJournal();
+            journal.RegisterPackageUse(thePackage, deploymentOne);
+
+            journal.ApplyRetention(PackageDirectory);
+
+            // TODO: Fix these assertions
+            // How do we ensure the given package will be marked for removal?
+            // How do we simulate needing more disk space
```
Comment (MissedTheMark) [inline source/Calamari.Tests/Fixtures/PackageRetention/JournalEntryFixture.cs:162]: You can use a real `IFreeSpaceChecker`, with a mock `IFileSystem` passed to it - and make the mock `fileSystem.GetDiskFreeSpace` call return the space you want it to.
Let me know if you want me to walk you through it.

File: source/Calamari.Tests/Fixtures/PackageRetention/JournalEntryFixture.cs [hunk 2]
```diff
@@ -128,12 +139,30 @@ public void WhenPackageIsRegisteredAndDeregistered_ThenUsageIsStillRecorded()
...
+            var thePackage = new PackageIdentity("Package", "1.0");
+            var deploymentOne = new ServerTaskId("Deployment-1");
+    
+            var journal = GetPackageJournal();
+            journal.RegisterPackageUse(thePackage, deploymentOne);
+
+            journal.ApplyRetention(PackageDirectory);
+
+            // TODO: Fix these assertions
+            // How do we ensure the given package will be marked for removal?
```
Comment (MissedTheMark) [inline source/Calamari.Tests/Fixtures/PackageRetention/JournalEntryFixture.cs:161]: Mock the retention algorithm to return the package you want to have removed.

**Review by MissedTheMark — CHANGES_REQUESTED** [review]
Couple of things to change, nothing huge.  Added some notes on how to get the free space tests to do what you need 👍

File: source/Calamari.Shared/Commands/ApplyDeltaCommand.cs [hunk 3]
```diff
@@ -54,7 +56,7 @@ public override int Execute(string[] commandLineArguments)
             try
             {
                 ValidateParameters(out basisFilePath, out deltaFilePath, out newFilePath);
-                freeSpaceChecker.EnsureDiskHasEnoughFreeSpace(PackageStore.GetPackagesDirectory());
+                packageJournal.ApplyRetention(PackageStore.GetPackagesDirectory());
```
[the remark this answers is not in the document]
Comment (desruc) [reply source/Calamari.Shared/Commands/ApplyDeltaCommand.cs:59]: > Given that we already have a check in `ApplyRetention`, perhaps we should add a dependency on the old FreeSpaceChecker there. Would make the calling code cleaner too if it's inside `ApplyRetention`

I've taken this approach.

`ApplyRetention` calls the old `EnsureDiskHasEnoughFreeSpace` method if package retention is disabled. I've also wrapped it in a try/catch.

**Review by MissedTheMark — APPROVED** [review]
👍

**Pull request merged**